[FLINK-2437] Fix default constructor detection in TypeExtractor.analyzePojo

ggevay opened the pull request:
- handle the case of a non-public default constructor
- added a missing "return null"
- removed some duplicating of the word "class" when printing class
  names, because class.toString also prepends it to the class name
- fixed a typo in the documentation describing POJOs

ggevay commented:
I removed the added "return null", I just realized that it is not needed.

gyfora commented:
Could you please add a test case that verifies the correct behaviour?

StephanEwen commented:
Why are you changing all messages in this pull request? They were not wrong before...

mxm commented:
Stephan is right. The changed messages just add noise to the pull request which makes it harder to review.

ggevay commented:
They print things like
`Class class malom.GameState must have a default constructor to be used as a POJO.`
If you think that the double "class" should be left the way it is, then I'm sorry for changing them. I have now removed the changes from the PR.

I have also added a test.

StephanEwen commented:
Removing the double "class" is actually not bad. Was hard to get this from the diff alone.

BTW: Simpler way may be to change the `aClass.toString()` call to `aClass.getCanonicalName()`.

ggevay commented:
OK, I re-added the changes to remove the double "class". (Note, that the 3rd bullet in the opening comment refers to this doubling.)

mxm commented on flink-java/src/main/java/org/apache/flink/api/java/typeutils/TypeExtractor.java:
```diff
@@ -1328,24 +1329,29 @@ else if(typeHierarchy.size() <= 1) {
 				return null;
 			}
 		}
+		if(defaultConstructor != null && (defaultConstructor.getModifiers() & Modifier.PUBLIC) == 0) {
```
`if(defaultConstructor != null && Modifier.isPublic(defaultConstructor.getModifiers())` seems to be more readable to me but your approach is fine too.

ggevay replied:
You are right, I have changed  it.

mxm commented:
Your changes look good to merge.

mxm commented:
Thanks for your contribution!
